### PR TITLE
Custom workflows plan

### PR DESCRIPTION
**Pull request by veerajsk**
PopulateBlackboard:

- Adds a PopulateBlackboard plan implementation in the alica_standard_library which populates all protected blackboard keys from the input key `data` which is assumed to be json formatted. The type is taken from the blackboard blueprint & the json is parsed to set the value. 

- i.e. it effectively does `blackboard.set("protected_key", YAML::Load(blackboard.get<std::string>("data")).as<type_from_blueprint>()` for every protected key.

- If a key is not present in the input `data` it is skipped

- adds a unit test for testing PopulateBlackboard

CustomWorkflowTutorial:

- The PopulateBlackboard implementation is used to demonstrate how to create custom workflows by adding a tutorial in turtlesim.
- The tutorial makes use of PopulateBlackboard to parse & populate input keys for sub plans from a task msg published on a topic

Engine changes:

- The Parser in blackboard.h now simply forwards string values instead of calling Load on them. This is required to pass `json` values as strings instead of parsing them again which was resulting in incorrect behaviour.
- Need to read the access type of a blackboard value into the blueprint
- Provide access to the blackboard blueprint via BasicPlan & BasicBehaviour

Note: Change seems big mainly because of plan changes i.e. due to added/modified pml & beh files etc.

### CODE REVIEW

File: alica_standard_library/include/PopulateBlackboard.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+private:
+    using KnownTypes = std::tuple<bool, int64_t, uint64_t, double, std::string>;
+    static constexpr std::array<const char*, 5> _knownTypes{"bool", "int64", "uint64", "double", "std::string"};
+
+    void set(const std::string& key, const std::string& type);
+
+    template <std::size_t... Is>
+    void setHelper(const std::string& key, std::size_t typeIndex, std::index_sequence<Is...>)
+    {
+        struct Dummy
```
Comment (Dale-Koenig) [inline alica_standard_library/include/PopulateBlackboard.h:33]: Can you comment to explain what this is doing?

**Comment by Dale-Koenig** [comment]
>i.e. it effectively does blackboard.set("protected_key", YAML::Load(blackboard.get<std::string>("data")).as<type_from_blueprint>() for every protected key.

We only want to do this for every key in the json I think.  It is possible there are protected keys that are not meant to be set from the json

**Comment by bjoernschroeder** [comment]
> We only want to do this for every key in the json I think. It is possible there are protected keys that are not meant to be set from the json

But all protected keys which are not found in the json file are skipped. I dont think there should be any case where we have the key as protected in the blueprint and the same key in the json file and we dont want to set it, right?

**Comment by veerajsk** [comment]
> > i.e. it effectively does blackboard.set("protected_key", YAML::Load(blackboard.getstd::string("data")).as<type_from_blueprint>() for every protected key.
> 
> We only want to do this for every key in the json I think. It is possible there are protected keys that are not meant to be set from the json

yeah, that's the reason that if the key is not found in the json, we skip over to the next key without raising any kind of error. Also I think we don't require to map every key in the json to some key on the blackboard either, so there is no error raised for that either

**Comment by veerajsk** [comment]
> > We only want to do this for every key in the json I think. It is possible there are protected keys that are not meant to be set from the json
> 
> But all protected keys which are not found in the json file are skipped. I dont think there should be any case where we have the key as protected in the blueprint and the same key in the json file and we dont want to set it, right?

so, a key will be set if it is in the json & in the blackboard & it is protected.

File: alica_standard_library/include/PopulateBlackboard.h (hunk 1 above)
Comment (veerajsk) [reply alica_standard_library/include/PopulateBlackboard.h:33]: this is just syntactic sugar to call `setIfIndex` for every index in the index_sequence.
The comma operator makes sure `setIfIndex` is called, however the value of the expression `(setIfIndex, 0)` itself is `0`, so basically `dummyFunc` will be called with a bunch of 0s

**Comment by Dale-Koenig** [comment]
Looks like some of the blackboard mappings in the tutorial are broken now

**Comment by veerajsk** [comment]
> Looks like some of the blackboard mappings in the tutorial are broken now

strange, I tested the changes as well. Not sure how this happened. I fixed it though & tested again

**Review by bjoernschroeder — APPROVED** [review]

**Review by Dale-Koenig — APPROVED** [review]

**Pull request merged**